File: hardware/airsensor/airsensor.go
// Sparkfun BME280 temperature/pressure/humidity sensor
// https://www.sparkfun.com/sparkfun-atmospheric-sensor-breakout-bme280-qwiic.html
// https://cdn.sparkfun.com/assets/e/7/3/b/1/BME280_Datasheet.pdf

package airsensor

import (
	"errors"
	"math"

	"github.com/vincent99/velocipi-go/hardware/i2c"
)

const (
	DEFAULT_ADDRESS    = 0x77
	CALIBRATION_A_REG  = 0x88
	CALIBRATION_B_REG  = 0xE1
	CALIBRATION_H1_REG = 0xA1
	DATA_REG           = 0xF7
	CHIP_ID_REG        = 0xD0 // Chip ID
	RESET_REG          = 0xE0
	CONFIG_HUM_RES     = 0xF2 // Humidity config
	CONFIG_MEAS_RES    = 0xF4 // Temp/Pressure config
	CONFIG_RES         = 0xF5 // Other config
)

type Calibration struct {
	T1 uint16
	T2 int16
	T3 int16
	P1 uint16
	P2 int16
	P3 int16
	P4 int16
	P5 int16
	P6 int16
	P7 int16
	P8 int16
	P9 int16
	H1 uint8
	H2 int16
	H3 uint8
	H4 int16
	H5 int16
	H6 int8
}

type AirSensor struct {
	iface             *i2c.I2C
	config            *Config
	calibration       *Calibration
	tFine             int32
	referencePressure float32
}

type RunMode byte

const (
	SLEEP  RunMode = 0b00
	FORCED RunMode = 0b01
	NORMAL RunMode = 0b11
)

type StandbyConfig byte

const (
	SB_1    StandbyConfig = 0b000
	SB_10   StandbyConfig = 0b110 // Yes, they're out of order.
	SB_20   StandbyConfig = 0b111
	SB_62   StandbyConfig = 0b001
	SB_125  StandbyConfig = 0b010
	SB_250  StandbyConfig = 0b011
	SB_500  StandbyConfig = 0b100
	SB_1000 StandbyConfig = 0b101
)

type FilterConfig byte

const (
	FILTER_OFF FilterConfig = 0b000
	FILTER_2   FilterConfig = 0b001
	FILTER_4   FilterConfig = 0b010
	FILTER_8   FilterConfig = 0b011
	FILTER_16  FilterConfig = 0b100
)

type OversampleConfig byte

const (
	SKIPPED OversampleConfig = 0b000
	OS_1    OversampleConfig = 0b001
	OS_2    OversampleConfig = 0b010
	OS_4    OversampleConfig = 0b011
	OS_8    OversampleConfig = 0b100
	OS_16   OversampleConfig = 0b101
)

type Config struct {
	Address uint8
	Device  string

	Mode               RunMode
	Standby            StandbyConfig
	Filter             FilterConfig
	TempOversample     OversampleConfig
	TempCorrectionC    float32
	PressureOversample OversampleConfig
	HumidityOversample OversampleConfig
}

type Reading struct {
	TempC          float32 `json:"tempC"`
	TempF          float32 `json:"tempF"`
	PressureInches float32 `json:"pressureInches"`
	PressureMeters float32 `json:"pressureMeters"`
	PressureFeet   float32 `json:"pressureFeet"`
	Humidity       float32 `json:"humidity"`
	DewpointC      float32 `json:"dewpointC"`
	DewpointF      float32 `json:"dewpointF"`
}

func NewAirSensor() (*AirSensor, error) {
	return NewAirSensorWithOptions(&Config{
		Mode:               NORMAL,
		Standby:            SB_1,
		Filter:             FILTER_2,
		TempOversample:     OS_16,
		PressureOversample: OS_16,
		HumidityOversample: OS_16,
		TempCorrectionC:    0,
	})
}

func NewAirSensorWithOptions(opt *Config) (*AirSensor, error) {
	address := opt.Address
	if address == 0 {
		address = DEFAULT_ADDRESS
	}

	iface, err := i2c.New(opt.Device, address)

	v := &AirSensor{
		iface:             iface,
		config:            opt,
		calibration:       &Calibration{},
		referencePressure: 101325.0,
		tFine:             0,
	}

	if err != nil {
		return v, err
	}

	return v, v.Init()
}

func (v *AirSensor) Init() error {
	if !v.IsConnected() {
		return errors.New("air sensor not found")
	}

	chipId, err := v.iface.ReadRegisterU8(CHIP_ID_REG)
	if err != nil {
		return err
	}

	if chipId != 0x58 && chipId != 0x60 {
		return errors.New("air sensor has unrecognized chip id")
	}

	a, err := v.iface.ReadRegister(CALIBRATION_A_REG, 26)
	if err != nil {
		return err
	}

	b, err := v.iface.ReadRegister(CALIBRATION_B_REG, 8)
	if err != nil {
		return err
	}

	h1, err := v.iface.ReadRegisterU8(CALIBRATION_H1_REG)
	if err != nil {
		return err
	}

	v.calibration = &Calibration{
		T1: uint16(uint16(a[1])<<8 | uint16(a[0])),
		T2: int16(int16(a[3])<<8 | int16(a[2])),
		T3: int16(int16(a[5])<<8 | int16(a[4])),

		P1: uint16(uint16(a[7])<<8 | uint16(a[6])),
		P2: int16(int16(a[9])<<8 | int16(a[8])),
		P3: int16(int16(a[11])<<8 | int16(a[10])),
		P4: int16(int16(a[13])<<8 | int16(a[12])),
		P5: int16(int16(a[15])<<8 | int16(a[14])),
		P6: int16(int16(a[17])<<8 | int16(a[16])),
		P7: int16(int16(a[19])<<8 | int16(a[18])),
		P8: int16(int16(a[21])<<8 | int16(a[20])),
		P9: int16(int16(a[23])<<8 | int16(a[22])),

		H1: h1,
		H2: int16(int16(b[1])<<8 | int16(b[0])),
		H3: uint8(b[2]),
		H4: int16(int16(b[3])<<4 | (int16(b[4]) & 0x0F)),
		H5: int16(int16(b[5])<<4 | (int16(b[4]) >> 4 & 0x0F)),
		H6: int8(b[6]),
	}

	//fmt.Printf("Calibration A: %x\n", a)
	//fmt.Printf("Calibration B: %x\n", b)
	//fmt.Println(v.calibration)

	return v.WriteConfig()
}

func (v *AirSensor) IsConnected() bool {
	var buf []byte
	_, err := v.iface.WriteBytes(buf)
	return err == nil
}

func (v *AirSensor) Reset() error {
	return v.iface.WriteRegisterU8(RESET_REG, 0xB6)
}

// --------

func (v *AirSensor) WriteConfig() error {
	err := v.SetMode(SLEEP)
	if err != nil {
		return err
	}

	hum, err := v.iface.ReadRegisterU8(CONFIG_HUM_RES)
	if err != nil {
		return err
	}

	hum = (hum & 0b11111000) | byte(v.config.HumidityOversample)
	cfg := byte(v.config.Standby)<<5 | byte(v.config.Filter)<<2
	meas := byte(v.config.TempOversample)<<5 | byte(v.config.PressureOversample)<<2 | byte(v.config.Mode)

	//fmt.Printf("Write %x: %08b\n", CONFIG_HUM_RES, hum)
	err = v.iface.WriteRegisterU8(CONFIG_HUM_RES, hum)
	if err != nil {
		return err
	}

	//fmt.Printf("Write %x: %08b\n", CONFIG_RES, cfg)
	err = v.iface.WriteRegisterU8(CONFIG_RES, cfg)
	if err != nil {
		return err
	}

	//fmt.Printf("Write %x: %08b\n", CONFIG_MEAS_RES, meas)
	return v.iface.WriteRegisterU8(CONFIG_MEAS_RES, meas)
}

func (v *AirSensor) GetMode() (RunMode, error) {
	cfg, err := v.iface.ReadRegisterU8(CONFIG_MEAS_RES)
	if err != nil {
		return SLEEP, err
	}

	return RunMode(cfg & 0b11), nil
}

func (v *AirSensor) SetMode(val RunMode) error {
	cfg, err := v.iface.ReadRegisterU8(CONFIG_MEAS_RES)
	if err != nil {
		return err
	}

	cfg = (cfg & 0b11111100) | (byte(val) << 2)
	return v.iface.WriteRegisterU8(CONFIG_MEAS_RES, cfg)
}

// --------

func (v *AirSensor) Read() (r *Reading, err error) {
	r = &Reading{}

	raw, err := v.iface.ReadRegister(DATA_REG, 8)
	if err != nil {
		return r, err
	}

	//fmt.Printf("Read: %x\n", raw)

	p := int32(raw[0])<<12 | int32(raw[1])<<4 | (int32(raw[2]) >> 4 & 0x0F)
	t := int32(raw[3])<<12 | int32(raw[4])<<4 | (int32(raw[5]) >> 4 & 0x0F)
	h := uint16(raw[6])<<8 | uint16(raw[7])

	t1 := (((t >> 3) - int32(v.calibration.T1<<1)) * int32(v.calibration.T2)) >> 11
	t2 := (((((t >> 4) - int32(v.calibration.T1)) * ((t >> 4) - int32(v.calibration.T1))) >> 12) * int32(v.calibration.T3)) >> 14
	v.tFine = t1 + t2

	celsius := float32((v.tFine*5+128)>>8)/100 + v.config.TempCorrectionC
	fahrenheit := (celsius*9)/5 + 32

	//fmt.Printf("Temp: %f / %f\n", celsius, fahrenheit)

	press := float32(0)

	var p1 int64 = int64(v.tFine) - 128000
	var p2 int64 = p1 * p1 * int64(v.calibration.P6)
	p2 = p2 + (int64(p1*int64(v.calibration.P5)) << 17)
	p2 = p2 + (int64(v.calibration.P4) << 35)
	p1 = ((p1 * p1 * int64(v.calibration.P3)) >> 8) + ((p1 * int64(v.calibration.P2)) << 12)
	p1 = ((1 << 47) + p1) * (int64(v.calibration.P1)) >> 33

	if p1 != 0 {
		var pA int64 = 1048576 - int64(p)
		pA = (((pA << 31) - p2) * 3125) / p1
		p1 = (int64(v.calibration.P9) * (pA >> 13) * (pA >> 13)) >> 25
		p2 = (int64(v.calibration.P8) * pA) >> 19
		pA = ((pA + p1 + p2) >> 8) + (int64(v.calibration.P7) << 4)
		press = float32(pA / 256.0)
	}

	inches := press / 3386.39
	meters := (-44330.77) * float32(math.Pow(float64(press/v.referencePressure), 0.190263)-1.0)
	feet := meters * 3.28084

	//fmt.Printf("Pressure: %f\" %fm / %fft\n", inches, meters, feet)

	var h1 int32 = (v.tFine - 76800)
	h1 = ((((int32(h) << 14) - (int32(v.calibration.H4) << 20) - (int32(v.calibration.H5) * h1)) + (16384)) >> 15) * (((((((h1*int32(v.calibration.H6))>>10)*(((h1*int32(v.calibration.H3))>>11)+(32768)))>>10)+(2097152))*int32(v.calibration.H2) + 8192) >> 14)
	h1 = (h1 - (((((h1 >> 15) * (h1 >> 15)) >> 7) * int32(v.calibration.H1)) >> 4))
	h1 = min(max(h1, 0), 419430400)

	humidity := float32(h1>>12) / 1024.0
	//fmt.Printf("Humidity: %f%%\n", humidity)

	ratio := 373.15 / (273.15 + float64(celsius))
	rhs := -7.90298 * (ratio - 1)
	rhs += 5.02808 * math.Log10(ratio)
	rhs += -1.3816e-7 * (math.Pow(10, (11.344*(1-1/ratio))) - 1)
	rhs += 8.1328e-3 * (math.Pow(10, (-3.49149*(ratio-1))) - 1)
	rhs += float64(math.Log10(1013.246))
	vp := math.Pow(10, rhs-3) * float64(humidity)
	th := math.Log(vp / 0.61078)

	dewpointCelsius := float32((241.88 * th) / (17.558 - th))
	dewpointFahrenheit := dewpointCelsius*9/5 + 32

	// fmt.Printf("Dewpoint: %f C / %f F\n", dewpointCelsius, dewpointFahrenheit)

	return &Reading{
		TempC:          celsius,
		TempF:          fahrenheit,
		PressureInches: inches,
		PressureMeters: meters,
		PressureFeet:   feet,
		Humidity:       float32(humidity),
		DewpointC:      dewpointCelsius,
		DewpointF:      dewpointFahrenheit,
	}, nil
}

File: src/hardware/air-sensor.ts
// Sparkfun BME280 temperature/pressure/humidity sensor
// https://www.sparkfun.com/sparkfun-atmospheric-sensor-breakout-bme280-qwiic.html
// https://cdn.sparkfun.com/assets/e/7/3/b/1/BME280_Datasheet.pdf

import I2C, { Peripheral } from "./i2c";

// --- Constants ---
const DEFAULT_ADDRESS = 0x77;
const CALIBRATION_A_REG = 0x88;
const CALIBRATION_B_REG = 0xE1;
const CALIBRATION_H1_REG = 0xA1;
const DATA_REG = 0xF7;
const CHIP_ID_REG = 0xD0; // Chip ID
const RESET_REG = 0xE0;
const CONFIG_HUM_RES = 0xF2; // Humidity config
const CONFIG_MEAS_RES = 0xF4; // Temp/Pressure config
const CONFIG_RES = 0xF5; // Other config

// --- Enums and Types ---

export enum RunMode {
  SLEEP = 0b00,
  FORCED = 0b01,
  NORMAL = 0b11,
}

export enum StandbyConfig {
  SB_1 = 0b000,
  SB_10 = 0b110, // Yes, they're out of order.
  SB_20 = 0b111,
  SB_62 = 0b001,
  SB_125 = 0b010,
  SB_250 = 0b011,
  SB_500 = 0b100,
  SB_1000 = 0b101,
}

export enum FilterConfig {
  FILTER_OFF = 0b000,
  FILTER_2 = 0b001,
  FILTER_4 = 0b010,
  FILTER_8 = 0b011,
  FILTER_16 = 0b100,
}

export enum OversampleConfig {
  SKIPPED = 0b000,
  OS_1 = 0b001,
  OS_2 = 0b010,
  OS_4 = 0b011,
  OS_8 = 0b100,
  OS_16 = 0b101,
}

export type Calibration = {
  T1: number; // uint16
  T2: number; // int16
  T3: number; // int16
  P1: number; // uint16
  P2: number; // int16
  P3: number; // int16
  P4: number; // int16
  P5: number; // int16
  P6: number; // int16
  P7: number; // int16
  P8: number; // int16
  P9: number; // int16
  H1: number; // uint8
  H2: number; // int16
  H3: number; // uint8
  H4: number; // int16
  H5: number; // int16
  H6: number; // int8
}

export type Config = {
  Address: number
  Mode: RunMode;
  Standby: StandbyConfig;
  Filter: FilterConfig;
  TempOversample: OversampleConfig;
  TempCorrectionC: number; // float32
  PressureOversample: OversampleConfig;
  HumidityOversample: OversampleConfig;
}

export type Reading = {
  tempC: number; // float32
  tempF: number; // float32
  pressureInches: number; // float32
  pressureMeters: number; // float32
  pressureFeet: number; // float32
  humidity: number; // float32
  dewpointC: number; // float32
  dewpointF: number; // float32
}

const defaultConfig: Config = {
  Mode: RunMode.NORMAL,
  Standby: StandbyConfig.SB_1,
  Filter: FilterConfig.FILTER_2,
  TempOversample: OversampleConfig.OS_16,
  PressureOversample: OversampleConfig.OS_16,
  HumidityOversample: OversampleConfig.OS_16,
  TempCorrectionC: 0,
  Address: DEFAULT_ADDRESS
}

// --- Main Class ---

export class AirSensor {
  private iface: Peripheral;
  private config: Config;
  private calibration: Calibration;
  private tFine: number; // int32 - compensated temperature intermediate value
  private referencePressure = 101325;

  constructor(bus: I2C, config?: Config) {
    this.config = { ...defaultConfig, ...config || {} }
    this.iface = bus.wrap(this.config.Address)
    this.calibration = {} as Calibration;
    this.tFine = 0;
    this.referencePressure = 101325.0;
  }

  public async init() {
    if (!this.iface?.isConnected()) {
      throw new Error("air sensor not found");
    }

    const chipId = await this.iface.readRegisterU8(CHIP_ID_REG);

    if (chipId !== 0x58 && chipId !== 0x60) {
      throw new Error(`air sensor has unrecognized chip id: 0x${chipId.toString(16)}`);
    }

    // Read calibration registers
    const a = await this.iface.readRegister(CALIBRATION_A_REG, 26);
    const b = await this.iface.readRegister(CALIBRATION_B_REG, 8);
    const h1 = await this.iface.readRegisterU8(CALIBRATION_H1_REG);

    // Parse calibration values (Little Endian, as per BME280 datasheet)
    this.calibration = {
      T1: (a[ 1 ] << 8) | a[ 0 ],
      T2: toSigned16((a[ 3 ] << 8) | a[ 2 ]),
      T3: toSigned16((a[ 5 ] << 8) | a[ 4 ]),

      P1: (a[ 7 ] << 8) | a[ 6 ],
      P2: toSigned16((a[ 9 ] << 8) | a[ 8 ]),
      P3: toSigned16((a[ 11 ] << 8) | a[ 10 ]),
      P4: toSigned16((a[ 13 ] << 8) | a[ 12 ]),
      P5: toSigned16((a[ 15 ] << 8) | a[ 14 ]),
      P6: toSigned16((a[ 17 ] << 8) | a[ 16 ]),
      P7: toSigned16((a[ 19 ] << 8) | a[ 18 ]),
      P8: toSigned16((a[ 21 ] << 8) | a[ 20 ]),
      P9: toSigned16((a[ 23 ] << 8) | a[ 22 ]),

      H1: h1,
      H2: toSigned16((b[ 1 ] << 8) | b[ 0 ]),
      H3: b[ 2 ],
      H4: toSigned16((b[ 3 ] << 4) | (b[ 4 ] & 0x0F)),
      H5: toSigned16((b[ 5 ] << 4) | ((b[ 4 ] >> 4) & 0x0F)),
      H6: b[ 6 ] > 127 ? b[ 6 ] - 256 : b[ 6 ], // Convert uint8 to int8
    };

    await this.writeConfig();
  }

  public async reset() {
    await this.iface.writeRegisterU8(RESET_REG, 0xB6);
  }

  public async setConfig(config: Config) {
    this.config = {...defaultConfig, ...config}
    return this.writeConfig()
  }

  // -------- Config Methods --------
  private async writeConfig() {
    // Set mode to sleep before writing configuration registers
    await this.setMode(RunMode.SLEEP);

    let hum = await this.iface.readRegisterU8(CONFIG_HUM_RES);

    // Set humidity oversampling (bits 2:0)
    hum = (hum & 0b11111000) | this.config.HumidityOversample;

    // Set standby (bits 7:5) and filter (bits 4:2)
    const cfg = (this.config.Standby << 5) | (this.config.Filter << 2);

    // Set temp oversample (bits 7:5), pressure oversample (bits 4:2), and mode (bits 1:0)
    const meas = (this.config.TempOversample << 5) | (this.config.PressureOversample << 2) | this.config.Mode;

    await this.iface.writeRegisterU8(CONFIG_HUM_RES, hum);
    await this.iface.writeRegisterU8(CONFIG_RES, cfg);
    await this.iface.writeRegisterU8(CONFIG_MEAS_RES, meas);
  }

  public async getMode(): Promise<RunMode> {
    const cfg = await this.iface.readRegisterU8(CONFIG_MEAS_RES);
    // Mode is bits 1:0
    return cfg & 0b11;
  }

  public async setMode(val: RunMode) {
    let cfg = await this.iface.readRegisterU8(CONFIG_MEAS_RES);
    // Clear old mode bits (1:0) and set new mode bits
    cfg = (cfg & 0b11111100) | val;
    await this.iface.writeRegisterU8(CONFIG_MEAS_RES, cfg);
  }

  // -------- Reading Method (Compensation Logic) --------

  public async read(): Promise<Reading> {
    const raw = await this.iface.readRegister(DATA_REG, 8);

    // Parse raw data (MSB-LSB-XLSB format)
    // Pressure: raw[0:2] (20-bit, bits 3:0 of raw[2] are LSBs)
    const p = (raw[ 0 ] << 12) | (raw[ 1 ] << 4) | (raw[ 2 ] >> 4);

    // Temperature: raw[3:5] (20-bit, bits 3:0 of raw[5] are LSBs)
    const t = (raw[ 3 ] << 12) | (raw[ 4 ] << 4) | (raw[ 5 ] >> 4);

    // Humidity: raw[6:7] (16-bit)
    const h = (raw[ 6 ] << 8) | raw[ 7 ];

    // --- 1. Temperature Compensation (t_fine calculation) ---
    const _t1 = (t >> 3) - (this.calibration.T1 << 1);
    const _t2 = (((t >> 4) - this.calibration.T1) * ((t >> 4) - this.calibration.T1)) >> 12;

    const t1 = (_t1 * this.calibration.T2) >> 11;
    const t2 = (_t2 * this.calibration.T3) >> 14;

    this.tFine = t1 + t2;

    const celsius = (((this.tFine * 5 + 128) >> 8) / 100) + this.config.TempCorrectionC;
    const fahrenheit = (celsius * 9) / 5 + 32;

    // --- 2. Pressure Compensation ---
    function shiftl(num: bigint|number, by: number) {
      return BigInt(num) << BigInt(by)
    }

    function shiftr(num: bigint|number, by: number) {
      return BigInt(num) >> BigInt(by)
    }

    let p1 = BigInt(this.tFine - 128000);
    let p2 = BigInt(p1 * p1 * BigInt(this.calibration.P6));

    p2 = p2 + shiftl(p1 * BigInt(this.calibration.P5), 17)
    p2 = p2 + shiftl(this.calibration.P4, 35);
    p1 = shiftr(p1 * p1 * BigInt(this.calibration.P3), 8) + shiftl(p1 * BigInt(this.calibration.P2), 12)
    p1 = shiftr((shiftl(1,47) + p1) * BigInt(this.calibration.P1), 33)

    let press = 0;
    if (p1 !== BigInt(0)) {
      let pA = BigInt(1048576 - p)
      pA = ((shiftl(pA, 31) - p2) * BigInt(3125)) / p1
      p1 = shiftr(BigInt(this.calibration.P9) * shiftr(pA, 13) * shiftr(pA, 13), 25)
      p2 = shiftr(BigInt(this.calibration.P8) * pA, 19)

      pA = shiftr(pA + p1 + p2, 8) + shiftl(this.calibration.P7, 4)
      press = parseInt(BigInt(pA / BigInt(256)).toString(), 10); // Pressure in Pa
    }

    // --- 3. Altitude and Unit Conversions ---

    const inches = press / 3386.39; // Pa to inHg

    // Altitude (meters) formula: H = 44330.77 * (1 - (P/P_ref)^(1/5.255)) - Uses 1/0.190263 ~ 5.255
    const meters = (-44330.77) * (Math.pow(press / this.referencePressure, 0.190263) - 1.0);
    const feet = meters * 3.28084;

    // --- 4. Humidity Compensation ---
    let h1 = this.tFine - 76800;

    h1 = (
      ((
        ((h << 14) - (this.calibration.H4 << 20) - (this.calibration.H5 * h1)) + 16384
      ) >> 15) * (
        ((
          (
            (((h1 * this.calibration.H6) >> 10) * (
              ((h1 * this.calibration.H3) >> 11) + 32768
            )) >> 10
          ) + 2097152
        ) * this.calibration.H2 + 8192
        ) >> 14
      )
    )

    h1 = h1 - (((((h1 >> 15) * (h1 >> 15)) >> 7) * this.calibration.H1) >> 4);

    // Clamp to 0-100% (with factor 1024.0)
    h1 = Math.min(Math.max(h1, 0), 419430400);

    const humidity = (h1 >> 12) / 1024.0; // %RH

    // --- 5. Dewpoint Calculation (Magnus formula approximation) ---

    const ratio = 373.15 / (273.15 + celsius);
    let rhs = -7.90298 * (ratio - 1);
    rhs += 5.02808 * Math.log10(ratio);
    rhs += -1.3816e-7 * (Math.pow(10, (11.344 * (1 - 1 / ratio))) - 1);
    rhs += 8.1328e-3 * (Math.pow(10, (-3.49149 * (ratio - 1))) - 1);
    rhs += Math.log10(1013.246);
    const vp = Math.pow(10, rhs - 3) * humidity; // Vapor Pressure
    const th = Math.log(vp / 0.61078);

    const dewpointCelsius = (241.88 * th) / (17.558 - th);
    const dewpointFahrenheit = (dewpointCelsius * 9) / 5 + 32;

    return {
      tempC: celsius,
      tempF: fahrenheit,
      pressureInches: inches,
      pressureMeters: meters,
      pressureFeet: feet,
      humidity: humidity,
      dewpointC: dewpointCelsius,
      dewpointF: dewpointFahrenheit,
    };
  }
}

function toSigned16(val: number): number {
  return (val & 0x8000) ? val - 0x10000 : val;
}
